### ISSUE TITLE
[CT-663] YAML anchor expansion failure in 1.2.0a1 / `main`

### ISSUE THREAD

**Issue by jtcohen6** [dbt-labs/dbt-core#5268]
Proactively identified regression in `1.2.0a1`. If I had to guess, it might be related to https://github.com/dbt-labs/dbt-core/pull/5146?

Here's a profile that uses [YAML anchors](https://support.atlassian.com/bitbucket-cloud/docs/yaml-anchors/) to DRY up duplicated code. This is valid YAML, per ["online YAML parser"](https://yaml-online-parser.appspot.com/?yaml=%23+profiles.yml%0Agarage-postgres%3A%0A++outputs%3A%0A++++dev%3A+%26garage-pg%0A++++++database%3A+jerco%0A++++++host%3A+localhost%0A++++++pass%3A+nopass%0A++++++port%3A+5432%0A++++++schema%3A+dbt_jcohen%0A++++++threads%3A+5%0A++++++type%3A+postgres%0A++++++user%3A+jerco%0A++++prod%3A%0A++++++%3C%3C%3A+*garage-pg%0A++++++schema%3A+analytics%0A++target%3A+dev&type=json):
```yml
# profiles.yml
garage-postgres:
  outputs:
    dev: &garage-pg
      database: jerco
      host: localhost
      pass: nopass
      port: 5432
      schema: dbt_jcohen
      threads: 5
      type: postgres
      user: jerco
    prod:
      <<: *garage-pg
      schema: analytics
  target: dev
```
When running v1.1.0, dbt is perfectly happy to expand the YAML anchor:
```
$ dbt debug
11:02:25  Running with dbt=1.1.0
dbt version: 1.1.0
python version: 3.9.12
python path: /Users/jerco/dev/scratch/testy/env/bin/python3.9
os info: macOS-12.3.1-x86_64-i386-64bit
Using profiles.yml file at /Users/jerco/.dbt/profiles.yml
Using dbt_project.yml file at /Users/jerco/dev/scratch/testy/dbt_project.yml

Configuration:
  profiles.yml file [OK found and valid]
  dbt_project.yml file [OK found and valid]

Required dependencies:
 - git [OK found]

Connection:
  host: localhost
  port: 5432
  user: jerco
  database: jerco
  schema: dbt_jcohen
  search_path: None
  keepalives_idle: 0
  sslmode: None
  Connection test: [OK connection ok]

All checks passed!
```
Running `dbt-core` installed from `main`, it's not:
```
$ dbt debug
11:02:38  Running with dbt=1.2.0-a1
dbt version: 1.2.0-a1
python version: 3.9.12
python path: /Users/jerco/dev/product/dbt-core/env/bin/python3.9
os info: macOS-12.3.1-x86_64-i386-64bit
Using profiles.yml file at /Users/jerco/.dbt/profiles.yml
Using dbt_project.yml file at /Users/jerco/dev/scratch/testy/dbt_project.yml

11:02:39  Encountered an error:
Runtime Error

  dbt encountered an error while trying to read your profiles.yml file.

  Runtime Error
    Syntax error near line 55
    ------------------------------
    52 |       type: postgres
    53 |       user: jerco
    54 |     prod:
    55 |       <<: *garage-pg
    56 |       schema: analytics
    57 |   target: dev
    58 | sandbox-redshift:

    Raw Error:
    ------------------------------
    could not determine a constructor for the tag 'tag:yaml.org,2002:merge'
      in "<unicode string>", line 55, column 7
```

**Comment by jtcohen6** [comment]
Decided this shouldn't actually be labeled `regression`, since it's yet to be released. But we definitely need a test case for this sort of thing, since folks do make use of yaml anchors! ([e.g.](https://discourse.getdbt.com/t/how-to-use-anchors-in-schema-yml/1156))

**Comment by jtcohen6** [comment]
Resolved by #5403. We ended up reverting #5146
